File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.cpp
class Trie{
public:
    Trie *chars[26];
    bool end = false;
};
class WordDictionary {
public:
    Trie *root;
    WordDictionary() {
        root = new Trie();
    }
    
    void addWord(string word) {
        Trie *node = root;
        
        for(char c:word){
            if(!node->chars[c-'a'])
                node->chars[c-'a'] = new Trie();
            node = node->chars[c-'a'];
        }
        node->end=true;
    }
    
    bool search(string word) {
        return searchHelper(word, root);
    }
    bool searchHelper(string sub, Trie *node){
        // cout<<sub<<endl;
        for(int j=0;j<sub.size();j++){
            if(sub[j]=='.'){
                bool found=false;
                for(int i=0;i<26;i++){
                    if(!found && node->chars[i])
                        found |= searchHelper(sub.substr(j+1), node->chars[i]);
                }
                return found;
            }
            else{
                if(!node->chars[sub[j] - 'a'])
                    return false;
                node = node->chars[sub[j] - 'a'];
            }
        }
        return node->end;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class Trienode:
    def __init__(self):
        self.child = {}
        self.end = False
        
class WordDictionary:

    def __init__(self):
        self.root = Trienode()

    def addWord(self, word: str) -> None:
        node = self.root
        
        for i in word:
            if i not in node.child:
                node.child[i] = Trienode()
            node = node.child[i]
            
        node.end = True

    def search(self, word: str) -> bool:
        def dfs(s, node):

            for i in range(s,len(word)):
                if word[i]!='.':
                    if word[i] not in node.child:
                        return False
                    node = node.child[word[i]]
                else:
                    for x in node.child.values():
                        if dfs(i+1, x):
                            return True
                    return False
            return node.end
        
        return dfs(0, self.root)
                
                


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)